### PR TITLE
Log the correct line-number when using SLF4j with netty if possible.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

At the moment we do not log the correct line number in many cases as it will log the line number of the logger wrapper itself. Slf4j does have an extra interface that can be used to filter out these nad make it more usable with logging wrappers.

Modifications:

Detect if the returned logger implements LocationAwareLogger and if so make use of its extra methods to be able to log the correct origin of the log request.

Result:

Better logging when using slf4j.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This is a reworked version of https://github.com/netty/netty/pull/7519.

File: common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+ * <a href="http://www.slf4j.org/">SLF4J</a> logger which is location aware and so will log the correct origin of the
+ * logging event by filter out the wrapper itself.
+ */
+final class LocationAwareSlf4JLogger extends AbstractInternalLogger {
+
+    // IMPORTANT: All our log methods first check if the log level is enabled before call the wrapped
+    // LocationAwareLogger.log(...) method. This is done to reduce GC creation that is caused by varargs.
+
+    private static final String FQCN = LocationAwareSlf4JLogger.class.getName();
+    private static final long serialVersionUID = -8292030083201538180L;
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java]: nit: could use 0 or 1 here (or your favorite number). Using the ugly id is only needed for compatibility with previous versions of the class before you added the `serialVersionUID`.

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java]: I just pressed "generate" :D

File: common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    @Override
+    public void info(String format, Object... argArray) {
+        if (isInfoEnabled()) {
+            log(INFO_INT, format, argArray);
+        }
+    }
+
+    @Override
+    public void info(String msg, Throwable t) {
```
Comment (carl-mastrangelo) [inline common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java]: t is unused?

File: common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    public void warn(String format, Object argA, Object argB) {
+        if (isWarnEnabled()) {
+            log(WARN_INT, format, argA, argB);
+        }
+    }
+
+    @Override
+    public void warn(String msg, Throwable t) {
+        if (isWarnEnabled()) {
+            log(WARN_INT, msg);
```
Comment (carl-mastrangelo) [inline common/src/main/java/io/netty/util/internal/logging/LocationAwareSlf4JLogger.java]: same here

File: common/src/main/java/io/netty/util/internal/logging/Slf4JLoggerFactory.java [hunk 4]
```diff
@@ -44,6 +46,12 @@ public Slf4JLoggerFactory() {
 
     @Override
     public InternalLogger newInstance(String name) {
-        return new Slf4JLogger(LoggerFactory.getLogger(name));
+        return wrapLogger(LoggerFactory.getLogger(name));
+    }
+
+    // package-private for testing.
+    static InternalLogger wrapLogger(Logger logger) {
+        return  logger instanceof LocationAwareLogger ?
```
Comment (carl-mastrangelo) [inline common/src/main/java/io/netty/util/internal/logging/Slf4JLoggerFactory.java]: `s/  / /`

**Comment by normanmaurer** [comment]
@carl-mastrangelo good catch! Addressed all your comments. PTAL again

**Pull request merged**